### ISSUE TITLE
path to firmware

### ISSUE THREAD

**Issue by jacobrosenthal** [spark/uber-library-example#3]
I already have a library live. My source and examples are in the root directory because people download and put in the Arduino libraries dir.

Thus I cant add a firmware subdirectory with my code. How about a location path in the spark.json for non standard paths?

**Comment by bblanchon** [comment]
Is there anything new here?

**Comment by joegoggins** [comment]
Hey @bblanchon , thanks for the ping.

Though we don't have anything slated for the current Sprint to address this, we have had a bunch of architectural brainstorming sessions speccing out the what the library system should look like in light of the [Spark CLI](https://github.com/spark/spark-cli) and [Spark Dev](https://www.spark.io/dev) (which did not exist when the initial library system was built). We will certainly accommodate the changes requested in this issue then.

Additionally, we went ahead and hired a really badass developer, Wojtek, who will likely lead this library system changes effort when it does happen. ( Ping @suda , welcome to this conversation, :) )

**Comment by suda** [comment]
Hi @bblanchon,

like Joe mentioned, it's not on this sprint, but we definitely will address this issue. Whole libraries system is due for an overhaul, including API endpoints for accessing libraries and supporting them in CLI and Dev. So it's definitely on our roadmap.

**Comment by bblanchon** [comment]
Excellent! Thanks.

**Comment by jacobrosenthal** [comment]
+1

**Comment by bblanchon** [comment]
Did you make any progress on this issue?

I see many users are still forking [ArduinoJson](https://github.com/bblanchon/ArduinoJson) just to create a `firmware/` folder and a `spark.json` 😞

**Comment by suda** [comment]
Yes, we're on the last stretch for releasing Libraries V2 :) We hope to ship them before the holidays.

**Comment by bblanchon** [comment]
I saw that you updated the format.
It looks like it's the same as Arduino IDE :+1:

[ArduinoJson](https://github.com/bblanchon/ArduinoJson) doesn't have `src/` folder, is that OK?
Also, it's a header only library, would that be a problem?

**Comment by suda** [comment]
Yes, `src` directory is required by Particle Libraries but for [Arduino 1.5 compatibility](https://github.com/arduino/Arduino/wiki/Arduino-IDE-1.5:-Library-specification#source-code) we would recommend using `src` directory anyway. Header only library should be fine

**Comment by bblanchon** [comment]
@suda Apparently it is required to have a `.cpp` file in that folder.
See https://github.com/bblanchon/ArduinoJson/pull/461
But it doesn't make sense for a header-only library...

It's been quite difficult to find a layout that suits all existing IDEs: Arduino 1 (which is still the default on Debian), Arduino 1.5, Stino, Visual Micro, PlatformIO, Sloeber...
Right now, I'm on a sweet spot where every IDE works fine... except Particle 😞 
I fear that adding a dummy `.cpp` file will start another crazy round of angry users 😨 

Can you remove this constraint?

**Comment by suda** [comment]
@bblanchon you're right. It has been queued up in: https://github.com/spark/particle-library-manager/issues/26
